File: client/lan_advertiser.py
#!/usr/bin/env python3
"""
Фейковый LAN-рекламщик для Minecraft PE 1.1.5 (протокол 113).

Как работает:
- Слушает UDP 19132
- Отвечает на Unconnected Ping (0x01 / 0x02) пакетом Unconnected Pong (0x1c)
- Minecraft видит «локальный мир» во вкладке Друзья / LAN

Важно:
- Работает, когда устройства в одной сети (обычный Wi-Fi или виртуальная сеть: ZeroTier, Radmin и т.п.)
- Сам по себе не прокидывает трафик через интернет — для этого позже добавим прокси/туннель
"""

import socket
import struct
import threading
import time
import random
from typing import Optional

# RakNet Magic
RAKNET_MAGIC = bytes([
    0x00, 0xff, 0xff, 0x00,
    0xfe, 0xfe, 0xfe, 0xfe,
    0xfd, 0xfd, 0xfd, 0xfd,
    0x12, 0x34, 0x56, 0x78,
])

# Протокол Minecraft PE 1.1.5
PROTOCOL_VERSION = 113
GAME_VERSION = "1.1.5"


def build_unconnected_pong(
    client_timestamp: int,
    server_guid: int,
    world_name: str,
    player_count: int = 1,
    max_players: int = 5,
    port: int = 19132,
) -> bytes:
    """Собирает пакет Unconnected Pong (0x1c) для PE 1.1.5"""
    # Строка MOTD в формате Bedrock/MCPE
    # MCPE;название;протокол;версия;онлайн;макс;guid;подзаголовок;режим;режим_num;порт4;порт6;
    motd = (
        f"MCPE;{world_name};{PROTOCOL_VERSION};{GAME_VERSION};"
        f"{player_count};{max_players};{server_guid};"
        f"{world_name};Survival;1;{port};{port};"
    )
    motd_bytes = motd.encode("utf-8")

    packet = bytearray()
    packet.append(0x1c)  # Unconnected Pong
    packet += struct.pack(">Q", client_timestamp)  # timestamp из пинга
    packet += struct.pack(">Q", server_guid)       # GUID сервера
    packet += RAKNET_MAGIC
    packet += struct.pack(">H", len(motd_bytes))   # длина строки
    packet += motd_bytes
    return bytes(packet)


class LANAdvertiser:
    """
    Слушает пинги Minecraft и отвечает фейковым миром.
    Запускается в отдельном потоке.
    """

    def __init__(
        self,
        world_name: str = "Multiplayer Snake",
        player_count: int = 1,
        max_players: int = 5,
        port: int = 19132,
        bind_ip: str = "0.0.0.0",
    ):
        self.world_name = world_name
        self.player_count = player_count
        self.max_players = max_players
        self.port = port
        self.bind_ip = bind_ip
        self.server_guid = random.getrandbits(64)
        self._sock: Optional[socket.socket] = None
        self._thread: Optional[threading.Thread] = None
        self._running = False
        self.pings_received = 0

    def start(self) -> bool:
        if self._running:
            return True
        try:
            self._sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            self._sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            # Чтобы можно было слушать на 19132 даже если Minecraft тоже пытается
            try:
                self._sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
            except (AttributeError, OSError):
                pass
            self._sock.bind((self.bind_ip, self.port))
            self._sock.settimeout(1.0)
            self._running = True
            self._thread = threading.Thread(target=self._loop, daemon=True)
            self._thread.start()
            return True
        except OSError as e:
            print(f"[LAN] Не удалось занять порт {self.port}: {e}")
            print("[LAN] Возможно, порт уже занят Minecraft или другим процессом.")
            return False

    def stop(self):
        self._running = False
        if self._sock:
            try:
                self._sock.close()
            except Exception:
                pass
            self._sock = None
        if self._thread:
            self._thread.join(timeout=2)
            self._thread = None

    def _loop(self):
        while self._running and self._sock:
            try:
                data, addr = self._sock.recvfrom(2048)
            except socket.timeout:
                continue
            except OSError:
                break

            if not data:
                continue

            packet_id = data[0]
            # 0x01 = Unconnected Ping, 0x02 = Unconnected Ping Open Connections
            if packet_id not in (0x01, 0x02):
                continue

            if len(data) < 25:  # минимум: id + timestamp(8) + magic(16)
                continue

            # timestamp — следующие 8 байт
            client_ts = struct.unpack(">Q", data[1:9])[0]
            self.pings_received += 1

            pong = build_unconnected_pong(
                client_timestamp=client_ts,
                server_guid=self.server_guid,
                world_name=self.world_name,
                player_count=self.player_count,
                max_players=self.max_players,
                port=self.port,
            )
            try:
                self._sock.sendto(pong, addr)
            except OSError:
                pass

    def update(self, world_name: str = None, player_count: int = None, max_players: int = None):
        if world_name is not None:
            self.world_name = world_name
        if player_count is not None:
            self.player_count = player_count
        if max_players is not None:
            self.max_players = max_players


def test_advertiser():
    """Быстрый тест — запускает рекламу на 30 секунд"""
    print("Запуск LAN-рекламщика для теста (30 сек)...")
    print("Открой Minecraft PE 1.1.5 → Играть → Друзья / LAN")
    adv = LANAdvertiser(world_name="§aТест Multiplayer Snake", player_count=1, max_players=5)
    if not adv.start():
        return
    print(f"Реклама запущена. GUID={adv.server_guid}")
    try:
        for i in range(30):
            time.sleep(1)
            if i % 5 == 0:
                print(f"  ... пингов получено: {adv.pings_received}")
    except KeyboardInterrupt:
        pass
    adv.stop()
    print("Остановлено.")


if __name__ == "__main__":
    test_advertiser()

File: android-app/app/src/main/java/com/snake/mcpe/LANAdvertiser.java
package com.snake.mcpe;

import android.util.Log;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.SocketException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.Random;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Fake LAN advertiser for Minecraft PE 1.1.5 (protocol 113).
 * Listens UDP 19132 and answers Unconnected Ping (0x01/0x02) with Unconnected Pong (0x1c).
 * This makes a world appear in Play → Friends / LAN.
 */
public class LANAdvertiser {
    private static final String TAG = "SnakeLAN";
    private static final int DEFAULT_PORT = 19132;
    private static final int PROTOCOL_VERSION = 113;
    private static final String GAME_VERSION = "1.1.5";

    // RakNet magic
    private static final byte[] RAKNET_MAGIC = new byte[] {
        0x00, (byte) 0xff, (byte) 0xff, 0x00,
        (byte) 0xfe, (byte) 0xfe, (byte) 0xfe, (byte) 0xfe,
        (byte) 0xfd, (byte) 0xfd, (byte) 0xfd, (byte) 0xfd,
        0x12, 0x34, 0x56, 0x78
    };

    private String worldName;
    private int playerCount;
    private int maxPlayers;
    private final int port;
    private final long serverGuid;
    private DatagramSocket socket;
    private Thread thread;
    private final AtomicBoolean running = new AtomicBoolean(false);
    private final AtomicInteger pingsReceived = new AtomicInteger(0);

    public LANAdvertiser(String worldName, int playerCount, int maxPlayers) {
        this(worldName, playerCount, maxPlayers, DEFAULT_PORT);
    }

    public LANAdvertiser(String worldName, int playerCount, int maxPlayers, int port) {
        this.worldName = worldName != null ? worldName : "Multiplayer Snake";
        this.playerCount = Math.max(1, playerCount);
        this.maxPlayers = Math.max(this.playerCount, maxPlayers);
        this.port = port > 0 ? port : DEFAULT_PORT;
        this.serverGuid = new Random().nextLong();
    }

    public synchronized boolean start() {
        if (running.get()) return true;
        try {
            socket = new DatagramSocket(null);
            socket.setReuseAddress(true);
            try {
                // Best-effort on devices that support it
                socket.setOption(java.net.StandardSocketOptions.SO_REUSEPORT, true);
            } catch (Exception ignored) {}
            socket.bind(new InetSocketAddress(port));
            socket.setSoTimeout(1000);
            running.set(true);
            thread = new Thread(this::loop, "SnakeLANAdvertiser");
            thread.setDaemon(true);
            thread.start();
            Log.i(TAG, "LAN advertiser started on " + port + " name=" + worldName);
            return true;
        } catch (SocketException e) {
            Log.e(TAG, "Cannot bind " + port + ": " + e.getMessage());
            closeSocketQuietly();
            running.set(false);
            return false;
        } catch (Exception e) {
            Log.e(TAG, "start failed", e);
            closeSocketQuietly();
            running.set(false);
            return false;
        }
    }

    public synchronized void stop() {
        running.set(false);
        closeSocketQuietly();
        if (thread != null) {
            try {
                thread.join(1500);
            } catch (InterruptedException ignored) {
                Thread.currentThread().interrupt();
            }
            thread = null;
        }
        Log.i(TAG, "LAN advertiser stopped. pings=" + pingsReceived.get());
    }

    public boolean isRunning() {
        return running.get();
    }

    public int getPingsReceived() {
        return pingsReceived.get();
    }

    public void update(String worldName, Integer playerCount, Integer maxPlayers) {
        if (worldName != null) this.worldName = worldName;
        if (playerCount != null) this.playerCount = Math.max(1, playerCount);
        if (maxPlayers != null) this.maxPlayers = Math.max(this.playerCount, maxPlayers);
    }

    private void loop() {
        byte[] buf = new byte[2048];
        while (running.get() && socket != null && !socket.isClosed()) {
            try {
                DatagramPacket packet = new DatagramPacket(buf, buf.length);
                socket.receive(packet);
                if (packet.getLength() < 1) continue;
                int packetId = buf[0] & 0xff;
                // 0x01 Unconnected Ping, 0x02 Unconnected Ping Open Connections
                if (packetId != 0x01 && packetId != 0x02) continue;
                if (packet.getLength() < 25) continue; // id + ts(8) + magic(16)

                long clientTs = ByteBuffer.wrap(buf, 1, 8).order(ByteOrder.BIG_ENDIAN).getLong();
                pingsReceived.incrementAndGet();

                byte[] pong = buildUnconnectedPong(clientTs);
                DatagramPacket out = new DatagramPacket(
                    pong, pong.length, packet.getAddress(), packet.getPort());
                socket.send(out);
            } catch (java.net.SocketTimeoutException ignored) {
                // normal
            } catch (Exception e) {
                if (running.get()) {
                    Log.w(TAG, "recv/send: " + e.getMessage());
                }
                break;
            }
        }
    }

    private byte[] buildUnconnectedPong(long clientTimestamp) {
        // MCPE;name;protocol;version;online;max;guid;subname;mode;mode_num;port4;port6;
        String motd = "MCPE;" + worldName + ";" + PROTOCOL_VERSION + ";" + GAME_VERSION + ";"
            + playerCount + ";" + maxPlayers + ";" + (serverGuid & 0x7fffffffffffffffL) + ";"
            + worldName + ";Survival;1;" + port + ";" + port + ";";
        byte[] motdBytes = motd.getBytes(StandardCharsets.UTF_8);

        ByteBuffer bb = ByteBuffer.allocate(1 + 8 + 8 + 16 + 2 + motdBytes.length);
        bb.order(ByteOrder.BIG_ENDIAN);
        bb.put((byte) 0x1c);                 // Unconnected Pong
        bb.putLong(clientTimestamp);
        bb.putLong(serverGuid);
        bb.put(RAKNET_MAGIC);
        bb.putShort((short) motdBytes.length);
        bb.put(motdBytes);
        return bb.array();
    }

    private void closeSocketQuietly() {
        if (socket != null) {
            try {
                socket.close();
            } catch (Exception ignored) {}
            socket = null;
        }
    }
}
